### PR TITLE
[#41] feat: store dictionary source files in indexedDB

### LINKED ISSUE #41: Store dictionary sources in LocalStorage or IndexedDB

### PR DESCRIPTION
**Pull request by jienius**
[Please delete all explanatory text in square brackets prior to creating the pull request!]
[Write a short explanation of your changes]
Added in IndexedDB Access library Dexie.js
Added in basic functionalities to store files inside IndexedDB

[Link to an existing issue number]
Fixes #41 
 OR
Resolves #

### CODE REVIEW

File: src/worker/main.ts [hunk 1]
```diff
@@ -1 +1,32 @@
...
+//   const payload = event.data as ISaveFileEventPayload;
+//   indexedDB
+//     .saveFile(payload.name, payload.file)
+//     .then(() => {
+//       postMessage("Save Success");
+//     })
+//     .catch(() => {
+//       postMessage("Save Failed");
+//     });
+// };
```
Comment (eddieantonio) [inline src/worker/main.ts]: Please do not commit commented out code — that's what the commit log is for!

```suggestion
```

File: src/worker/main.ts [hunk 2]
```diff
@@ -1 +1,32 @@
-self.postMessage("hello, from worker");
+import indexedDB from "./indexedDBAccess";
+
+export interface ISaveFileEventPayload {
```
Comment (eddieantonio) [inline src/worker/main.ts]: Redundant `I` in the name.

```suggestion
export interface SaveFileEventPayload {
```

File: src/worker/indexedDBAccess.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+const FILES_TABLE_NAME = "files";
+const FILES_TABLE_SCHEMA = "++id, name, file";
+
+interface IFile {
+  id?: number;
+  name: string;
+  file: File;
+}
+
+class IndexDBAccess extends Dexie {
```
Comment (eddieantonio) [inline src/worker/indexedDBAccess.ts:25]: A quick look through [Dexie's documentation](https://dexie.org/docs/Tutorial/Best-Practices) doesn't show using it by extending or using inheritence;

What advanatages does inheritence here have over composition (e.g., just using the `db` as a property)?

File: src/worker/indexedDBAccess.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * +-------+
+ * | name  |
+ * +-------+
+ * | file  |
+ * +-------+
+ */
+const FILES_TABLE_NAME = "files";
+const FILES_TABLE_SCHEMA = "++id, name, file";
+
+interface IFile {
```
Comment (eddieantonio) [inline src/worker/indexedDBAccess.ts]:
```suggestion
interface StoredFile {
```

File: src/worker/main.ts (hunk 2 above)
Comment (eddieantonio) [reply src/worker/main.ts]: Is this the same interface as in `indexedDBAccess`? Can these two interface definitions be merged?

File: src/worker/indexedDBAccess.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  file: File;
+}
+
+class IndexDBAccess extends Dexie {
+  files: Dexie.Table<IFile, number>;
+
+  constructor() {
+    super(DB_NAME);
+    this.version(VERSION).stores({ files: FILES_TABLE_SCHEMA });
+    this.files = this.table(FILES_TABLE_NAME);
```
Comment (eddieantonio) [inline src/worker/indexedDBAccess.ts:31]: Doesn't Dexie export all tables as properties like this anyway? e.g., can't you do `db.files.where(...)` or `db.files.put(...)` etc?

File: src/worker/main.ts [hunk 6]
```diff
@@ -1 +1,32 @@
...
+  name: string;
+  file: File;
+}
+
+const handleSaveFileEvent = (event: MessageEvent) => {
+  const payload = event.data as ISaveFileEventPayload;
```
Comment (eddieantonio) [inline src/worker/main.ts]:
```suggestion
  const payload = event.data as SaveFileEventPayload;
```

File: src/worker/indexedDBAccess.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+const FILES_TABLE_SCHEMA = "++id, name, file";
+
+interface IFile {
+  id?: number;
+  name: string;
+  file: File;
+}
+
+class IndexDBAccess extends Dexie {
+  files: Dexie.Table<IFile, number>;
```
Comment (eddieantonio) [inline src/worker/indexedDBAccess.ts]:
```suggestion
  files: Dexie.Table<StoredFile, number>;
```

File: src/worker/indexedDBAccess.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * files Table Scehma
+ * +-------+
+ * | id    |
+ * +-------+
+ * | name  |
+ * +-------+
+ * | file  |
+ * +-------+
```
Comment (eddieantonio) [inline src/worker/indexedDBAccess.ts]: Good diagram! Could you mark which one is the primary key (e.g., indexed)? I believe `id` is an autoincrement primary key, right?

**Review by eddieantonio — CHANGES_REQUESTED** [review]
Good job finding Dexie! It seems really neat, and seems to provide a clean, intuitive interface to IndexedDB. That said, I have some concerns! 😂

Note: I know that IndexedDB itself likes declaring interfaces with an `I` prefix, but I think it is completely unnecessary in our own custom interfaces in TypeScript. After all, if the `I` is to remind us that it's an interface, the IDE and compiler can tell us that instead!

File: src/worker/indexedDBAccess.ts (hunk 3 above)
Comment (jienius) [reply src/worker/indexedDBAccess.ts:25]: I followed their recommendation when using Dexie with TypeScript: https://dexie.org/docs/Typescript
It provides better type safety and simpler code

Comment (eddieantonio) [reply src/worker/indexedDBAccess.ts:25]: Ah, okay, makes sense! That functionality _is_ already provided, but we would just like TypeScript to know exactly what happens when we access the `db.files` attribute!

File: src/worker/indexedDBAccess.ts (hunk 5 above)
Comment (eddieantonio) [reply src/worker/indexedDBAccess.ts:31]: Ignore this! I got it :)

File: src/worker/main.ts (hunk 2 above)
Comment (jienius) [reply src/worker/main.ts]: You can, but I don't want them to share the same interface as one of them is for DB access while the other one is related to UI event

File: src/worker/indexedDBAccess.ts (hunk 8 above)
Comment (jienius) [reply src/worker/indexedDBAccess.ts]: Yes

**Review by eddieantonio — APPROVED** [review]
LGTM

**Pull request merged**